pkg/util/tests/runtime.go

```diff
@@ -0,0 +1,126 @@
+package tests
+
+import (
+	"net"
+	"os"
+	"path/filepath"
+
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/k3s/pkg/daemons/control/deps"
+	"github.com/rancher/k3s/pkg/version"
+)
+
+// GenerateTestDataDir creates a temporary directory at "/tmp/k3s/<RANDOM_STRING>/".
+// The latest directory created with this function is soft linked to "/tmp/k3s/latest/".
+// This allows tests to replicate the "/var/lib/rancher/k3s" directory structure
+func GenerateTestDataDir(cnf *config.Control) error {
+	if err := os.MkdirAll(cnf.DataDir, 0700); err != nil {
+		return err
+	}
+	testDir, err := os.MkdirTemp(cnf.DataDir, "*")
+	if err != nil {
+		return err
+	}
+	// Remove old symlink and add new one
+	os.Remove(filepath.Join(cnf.DataDir, "latest"))
+	if err = os.Symlink(testDir, filepath.Join(cnf.DataDir, "latest")); err != nil {
+		return err
+	}
+	cnf.DataDir = testDir
+	cnf.DataDir, err = filepath.Abs(cnf.DataDir)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// CleanupTestDataDir removes the associated "/tmp/k3s/<RANDOM_STRING>"
+// directory.
+func CleanupTestDataDir(cnf *config.Control) {
+	os.RemoveAll(cnf.DataDir)
+}
+
+// GenerateTestRuntime creates a temporary data dir and configures
+// config.ControlRuntime with all the appropriate certificate keys.
+func GenerateTestRuntime(cnf *config.Control) error {
+
+	// Setup defaults for the config
+	_, clusterIPNet, _ := net.ParseCIDR("10.42.0.0/16")
+	cnf.ClusterIPRange = clusterIPNet
+
+	_, serviceIPNet, _ := net.ParseCIDR("10.43.0.0/16")
+	cnf.ServiceIPRange = serviceIPNet
+	cnf.ClusterDNS = net.ParseIP("10.43.0.10")
+	cnf.AdvertisePort = cnf.HTTPSPort
+
+	runtime := &config.ControlRuntime{}
+	if err := GenerateTestDataDir(cnf); err != nil {
+		return err
+	}
+
+	os.MkdirAll(filepath.Join(cnf.DataDir, "tls"), 0700)
+	os.MkdirAll(filepath.Join(cnf.DataDir, "cred"), 0700)
+
+	runtime.ClientCA = filepath.Join(cnf.DataDir, "tls", "client-ca.crt")
+	runtime.ClientCAKey = filepath.Join(cnf.DataDir, "tls", "client-ca.key")
```

Comment:
This all looks like it was copied from https://github.com/k3s-io/k3s/blob/f5fbb9a9a84dd758cfa2212167229e657e18c0c5/pkg/daemons/control/server.go#L234 - should we perhaps just move that into a library function?

Reply:
There are several things we do not want to happen in the testing infra, such as boostraping and starting the cluster. However, I will move all the `filepath.join` lines into a seperate function that can be used by both server.go and runtime.go.

Reply:
👍🏻  on that. I think the mkdir and path building stuff is worth sharing; certainly don't need to actually boostrap the cluster.